File: crates/ziwei_core/src/placement.rs
//! 本命安宫与安星的纯计算规则。

use super::{
    branch::Branch,
    five_element_bureau::FiveElementBureau,
    palace::PalaceName,
    position::{branch_from_yin0, branch_index_to_yin0},
    star::StarKey,
    stem::Stem,
};

/// 命宫与身宫地支。
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(crate) struct MingShenBranches {
    pub(crate) ming_palace: Branch,
    pub(crate) shen_palace: Branch,
}

/// 尚未附加星曜与四化关系的宫位落位。
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(crate) struct PalacePlacement {
    pub(crate) name: PalaceName,
    pub(crate) branch: Branch,
    pub(crate) stem: Stem,
}

/// 首批四颗辅星的落宫地支。
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(crate) struct MinorStarBranches {
    zuo_fu: Branch,
    you_bi: Branch,
    wen_chang: Branch,
    wen_qu: Branch,
}

/// 寅起正月；命宫逆时、身宫顺时。
pub(crate) fn compute_ming_shen_branches(month: u8, hour: u8) -> MingShenBranches {
    let ming_palace_yin0 = (i32::from(month) - i32::from(hour)).rem_euclid(12) as u8;
    let shen_palace_yin0 = (i32::from(month) + i32::from(hour)).rem_euclid(12) as u8;
    MingShenBranches {
        ming_palace: branch_from_yin0(ming_palace_yin0),
        shen_palace: branch_from_yin0(shen_palace_yin0),
    }
}

/// 五虎遁：按子为零的地支下标生成十二宫干。
pub(crate) fn compute_palace_stems(birth_stem: Stem) -> [Stem; 12] {
    let yin_head_stem_index = birth_stem.yin_head_stem().index() as u8;
    std::array::from_fn(|branch_index| {
        let branch_index = u8::try_from(branch_index).expect("twelve branches fit in u8");
        let yin0 = branch_index_to_yin0(branch_index);
        Stem::from_index((yin_head_stem_index + yin0).rem_euclid(10))
    })
}

/// 命宫干支确定五行局。
pub(crate) fn bureau_from_ming_palace(
    ming_palace_branch: Branch,
    palace_stems_by_branch: &[Stem; 12],
) -> FiveElementBureau {
    FiveElementBureau::from_ming_palace(
        palace_stems_by_branch[ming_palace_branch.index()],
        ming_palace_branch,
    )
}

/// 按地支下标（子为零）生成十二宫名、支、干坐标。
pub(crate) fn compute_palace_placements(
    ming_palace_branch: Branch,
    palace_stems_by_branch: &[Stem; 12],
) -> [PalacePlacement; 12] {
    let placeholder = PalacePlacement {
        name: PalaceName::Ming,
        branch: Branch::Zi,
        stem: Stem::Jia,
    };
    let mut placements_by_branch = [placeholder; 12];

    for name in PalaceName::ALL {
        let branch_index = (ming_palace_branch.index() as i32
            - i32::try_from(name.index()).expect("palace name index fits in i32"))
        .rem_euclid(12) as u8;
        let branch = Branch::from_index(branch_index);
        placements_by_branch[usize::from(branch_index)] = PalacePlacement {
            name,
            branch,
            stem: palace_stems_by_branch[usize::from(branch_index)],
        };
    }

    placements_by_branch
}

/// 计算首批四颗辅星的落宫地支。
pub(crate) fn compute_minor_star_branches(month: u8, hour: u8) -> MinorStarBranches {
    MinorStarBranches {
        zuo_fu: branch_from_yin0((2 + i32::from(month)).rem_euclid(12) as u8),
        you_bi: branch_from_yin0((8 - i32::from(month)).rem_euclid(12) as u8),
        wen_chang: branch_from_yin0((8 - i32::from(hour)).rem_euclid(12) as u8),
        wen_qu: branch_from_yin0((2 + i32::from(hour)).rem_euclid(12) as u8),
    }
}

/// 计算十四主星的落宫地支。
pub(crate) fn compute_major_star_branches(day: u8, bureau_number: u8) -> [Branch; 18] {
    let bureau_number = i32::from(bureau_number);
    let day = i32::from(day);
    let ceiling_quotient = (day + bureau_number - 1) / bureau_number;
    let shortfall = ceiling_quotient * bureau_number - day;
    let signed_shortfall = match shortfall {
        0 => 0,
        value if value.rem_euclid(2) == 1 => -value,
        value => value,
    };
    let zi_wei_yin0 = ((ceiling_quotient - 1) + signed_shortfall).rem_euclid(12) as u8;
    let tian_fu_yin0 = (-i32::from(zi_wei_yin0)).rem_euclid(12) as u8;

    let mut branches_by_star = [Branch::Zi; 18];
    for (star_key, yin0_offset) in [
        (StarKey::ZiWei, 0),
        (StarKey::TianJi, 1),
        (StarKey::TaiYang, 3),
        (StarKey::WuQu, 4),
        (StarKey::TianTong, 5),
        (StarKey::LianZhen, 8),
    ] {
        set_star_branch(
            &mut branches_by_star,
            star_key,
            (i32::from(zi_wei_yin0) - yin0_offset).rem_euclid(12) as u8,
        );
    }
    for (star_key, yin0_offset) in [
        (StarKey::TianFu, 0),
        (StarKey::TaiYin, 1),
        (StarKey::TanLang, 2),
        (StarKey::JuMen, 3),
        (StarKey::TianXiang, 4),
        (StarKey::TianLiang, 5),
        (StarKey::QiSha, 6),
        (StarKey::PoJun, 10),
    ] {
        set_star_branch(
            &mut branches_by_star,
            star_key,
            (i32::from(tian_fu_yin0) + yin0_offset).rem_euclid(12) as u8,
        );
    }

    branches_by_star
}

/// 把首批四颗辅星合并到十八星落宫表。
pub(crate) fn merge_star_branches(
    mut branches_by_star: [Branch; 18],
    minor_star_branches: MinorStarBranches,
) -> [Branch; 18] {
    branches_by_star[StarKey::ZuoFu.index()] = minor_star_branches.zuo_fu;
    branches_by_star[StarKey::YouBi.index()] = minor_star_branches.you_bi;
    branches_by_star[StarKey::WenChang.index()] = minor_star_branches.wen_chang;
    branches_by_star[StarKey::WenQu.index()] = minor_star_branches.wen_qu;
    branches_by_star
}

const fn set_star_branch(branches_by_star: &mut [Branch; 18], star_key: StarKey, yin0: u8) {
    branches_by_star[star_key.index()] = branch_from_yin0(yin0);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn palace_stems_follow_five_tiger_escape_for_every_year_stem_and_branch() {
        let branches_from_yin_to_chou = [
            Branch::Yin,
            Branch::Mao,
            Branch::Chen,
            Branch::Si,
            Branch::Wu,
            Branch::Wei,
            Branch::Shen,
            Branch::You,
            Branch::Xu,
            Branch::Hai,
            Branch::Zi,
            Branch::Chou,
        ];
        let five_tiger_cases = [
            (
                [Stem::Jia, Stem::Ji],
                [
                    Stem::Bing,
                    Stem::Ding,
                    Stem::Wu,
                    Stem::Ji,
                    Stem::Geng,
                    Stem::Xin,
                    Stem::Ren,
                    Stem::Gui,
                    Stem::Jia,
                    Stem::Yi,
                    Stem::Bing,
                    Stem::Ding,
                ],
            ),
            (
                [Stem::Yi, Stem::Geng],
                [
                    Stem::Wu,
                    Stem::Ji,
                    Stem::Geng,
                    Stem::Xin,
                    Stem::Ren,
                    Stem::Gui,
                    Stem::Jia,
                    Stem::Yi,
                    Stem::Bing,
                    Stem::Ding,
                    Stem::Wu,
                    Stem::Ji,
                ],
            ),
            (
                [Stem::Bing, Stem::Xin],
                [
                    Stem::Geng,
                    Stem::Xin,
                    Stem::Ren,
                    Stem::Gui,
                    Stem::Jia,
                    Stem::Yi,
                    Stem::Bing,
                    Stem::Ding,
                    Stem::Wu,
                    Stem::Ji,
                    Stem::Geng,
                    Stem::Xin,
                ],
            ),
            (
                [Stem::Ding, Stem::Ren],
                [
                    Stem::Ren,
                    Stem::Gui,
                    Stem::Jia,
                    Stem::Yi,
                    Stem::Bing,
                    Stem::Ding,
                    Stem::Wu,
                    Stem::Ji,
                    Stem::Geng,
                    Stem::Xin,
                    Stem::Ren,
                    Stem::Gui,
                ],
            ),
            (
                [Stem::Wu, Stem::Gui],
                [
                    Stem::Jia,
                    Stem::Yi,
                    Stem::Bing,
                    Stem::Ding,
                    Stem::Wu,
                    Stem::Ji,
                    Stem::Geng,
                    Stem::Xin,
                    Stem::Ren,
                    Stem::Gui,
                    Stem::Jia,
                    Stem::Yi,
                ],
            ),
        ];

        for (birth_stems, expected_stems_from_yin_to_chou) in five_tiger_cases {
            for birth_stem in birth_stems {
                let actual_stems_by_branch = compute_palace_stems(birth_stem);

                for (branch, expected_stem) in branches_from_yin_to_chou
                    .into_iter()
                    .zip(expected_stems_from_yin_to_chou)
                {
                    assert_eq!(
                        actual_stems_by_branch[branch.index()],
                        expected_stem,
                        "birth_stem={birth_stem:?} branch={branch:?}"
                    );
                }
            }
        }
    }

    #[test]
    fn tian_fu_mirrors_zi_wei_for_every_supported_day_and_bureau() {
        for day in 1..=30u8 {
            for bureau_number in [2u8, 3, 4, 5, 6] {
                let major_star_branches = compute_major_star_branches(day, bureau_number);
                let zi_wei_yin0 = branch_index_to_yin0(
                    u8::try_from(major_star_branches[StarKey::ZiWei.index()].index())
                        .expect("branch index fits in u8"),
                );
                let tian_fu_yin0 = branch_index_to_yin0(
                    u8::try_from(major_star_branches[StarKey::TianFu.index()].index())
                        .expect("branch index fits in u8"),
                );
                assert_eq!(
                    tian_fu_yin0,
                    (-i32::from(zi_wei_yin0)).rem_euclid(12) as u8,
                    "day={day} bureau={bureau_number}"
                );
            }
        }
    }

    #[test]
    fn palace_placements_include_each_name_and_branch_once() {
        let palace_stems_by_branch = compute_palace_stems(Stem::Jia);
        let placements_by_branch = compute_palace_placements(Branch::Zi, &palace_stems_by_branch);

        for name in PalaceName::ALL {
            assert_eq!(
                placements_by_branch
                    .iter()
                    .filter(|placement| placement.name == name)
                    .count(),
                1
            );
        }
        for branch_index in 0..12u8 {
            assert_eq!(
                placements_by_branch[usize::from(branch_index)]
                    .branch
                    .index(),
                usize::from(branch_index)
            );
        }
    }

    #[test]
    fn ming_shen_branches_follow_all_month_hour_formulas() {
        for month in 0..12u8 {
            for hour in 0..12u8 {
                let ming_shen = compute_ming_shen_branches(month, hour);

                assert_eq!(
                    ming_shen.ming_palace,
                    branch_from_yin0((i32::from(month) - i32::from(hour)).rem_euclid(12) as u8)
                );
                assert_eq!(
                    ming_shen.shen_palace,
                    branch_from_yin0((i32::from(month) + i32::from(hour)).rem_euclid(12) as u8)
                );
            }
        }
    }

    #[test]
    fn minor_star_branches_follow_all_month_hour_formulas() {
        for month in 0..12u8 {
            for hour in 0..12u8 {
                let minor_star_branches = compute_minor_star_branches(month, hour);

                assert_eq!(
                    minor_star_branches.zuo_fu,
                    branch_from_yin0((2 + i32::from(month)).rem_euclid(12) as u8)
                );
                assert_eq!(
                    minor_star_branches.you_bi,
                    branch_from_yin0((8 - i32::from(month)).rem_euclid(12) as u8)
                );
                assert_eq!(
                    minor_star_branches.wen_chang,
                    branch_from_yin0((8 - i32::from(hour)).rem_euclid(12) as u8)
                );
                assert_eq!(
                    minor_star_branches.wen_qu,
                    branch_from_yin0((2 + i32::from(hour)).rem_euclid(12) as u8)
                );
            }
        }
    }
}

File: src/models/placement.zig
//! 本命安宫、安星、四化与十二宫图的纯计算规则。
//!
//! 规则迁自 `backup-main-rust-before-reset-2026-08-13` 的安宫、安星、四化与
//! 本命盘装配实现；独立落宫期望位于 `natal.zig`，来源是该分支
//! `natal_placement_examples.rs` 标注的 Issues #267、#268、#269、#272。
//! 适用口径为这些已确认命例采用的项目规则，不作单一外部流派声明。

const std = @import("std");
const primitive = @import("primitive.zig");
const five_element_bureau = @import("five_element_bureau.zig");
const palace = @import("palace.zig");
const star = @import("star.zig");
const Transformation = @import("transformation.zig").Transformation;

const Branch = primitive.Branch;
const Stem = primitive.Stem;
const FiveElementBureau = five_element_bureau.FiveElementBureau;
const Palace = palace.Palace;
const PalaceName = palace.PalaceName;
const PalaceStars = palace.PalaceStars;
const StarName = star.StarName;

const zi_wei_yin0_by_bureau_day = buildZiWeiYin0ByBureauDay();
const major_branches_by_zi_wei_yin0 = buildMajorBranchesByZiWeiYin0();
const zuo_you_by_month = buildZuoYouByMonth();
const chang_qu_by_hour = buildChangQuByHour();
const cycle_len = Branch.all.len;

/// 宫位存储与安星公式共用的寅起坐标；子起 `Branch` 只在对外字段出现。
const YinIndex = enum(u4) {
    _,

    fn wrap(value: i32) YinIndex {
        return @enumFromInt(floorMod(value, cycle_len));
    }

    fn fromBranch(branch: Branch) YinIndex {
        return wrap(@as(i32, @intCast(branch.index())) - @as(i32, @intCast(Branch.yin.index())));
    }

    fn toInt(self: YinIndex) usize {
        return @intFromEnum(self);
    }

    fn toBranch(self: YinIndex) Branch {
        return primitive.branchFromIndex((self.toInt() + Branch.yin.index()) % cycle_len);
    }
};

/// 生年干与农历月、日、时辰确定的完整空间落位。
///
/// 返回值独立拥有全部宫位数据，不分配内存，也不保留输入借用。
pub const Layout = struct {
    palaces: [Branch.all.len]Palace,
    ming_palace_branch: Branch,
    shen_palace_name: PalaceName,
    shen_palace_branch: Branch,
    origin_palace_name: PalaceName,
    origin_palace_branch: Branch,
    bureau: FiveElementBureau,
};

/// 计算十二宫、星曜、四化、命宫、身宫、来因宫及五行局。
///
/// 农历月、日、时辰必须已经通过公开输入校验；本函数不分配内存，返回值
/// 独立拥有全部结果，可安全重入。
pub fn compute(birth_stem: Stem, month: u8, day: u8, hour: u8) Layout {
    const ming_palace = computeMingPalace(month, hour);
    const shen_palace = computeShenPalace(month, hour);
    const palace_stems = computePalaceStems(birth_stem);
    const bureau = bureauFromMingPalace(ming_palace, &palace_stems);
    const placements = computePalacePlacements(ming_palace, &palace_stems);
    const branches_by_star = computeStarBranches(bureau, month, day, hour);
    const origin_palace = YinIndex.fromBranch(primitive.originPalaceBranch(birth_stem));

    return .{
        .palaces = buildPalaces(origin_palace, &placements, &branches_by_star),
        .ming_palace_branch = ming_palace.toBranch(),
        .shen_palace_name = palaceNameAt(shen_palace, &placements),
        .shen_palace_branch = shen_palace.toBranch(),
        .origin_palace_name = palaceNameAt(origin_palace, &placements),
        .origin_palace_branch = origin_palace.toBranch(),
        .bureau = bureau,
    };
}

fn floorMod(value: i32, modulus: usize) usize {
    return @intCast(@mod(value, @as(i32, @intCast(modulus))));
}

/// 尚未附加星曜与四化关系的宫位落位。
const PalacePlacement = struct {
    name: PalaceName,
    branch: Branch,
    stem: Stem,
};

/// 寅起正月，逆时安命宫。
fn computeMingPalace(month: u8, hour: u8) YinIndex {
    return YinIndex.wrap(@as(i32, month) - @as(i32, hour));
}

/// 寅起正月，顺时安身宫。
fn computeShenPalace(month: u8, hour: u8) YinIndex {
    return YinIndex.wrap(@as(i32, month) + @as(i32, hour));
}

/// 五虎遁：按寅起坐标生成十二宫干。
fn computePalaceStems(birth_stem: Stem) [cycle_len]Stem {
    const yin_head_stem_index = primitive.stemIndex(primitive.yinHeadStem(birth_stem));
    var stems: [cycle_len]Stem = undefined;
    for (&stems, 0..) |*stem, yin0| {
        stem.* = primitive.stemFromIndex(yin_head_stem_index + yin0);
    }
    return stems;
}

/// 命宫干支确定五行局。
fn bureauFromMingPalace(
    ming_palace: YinIndex,
    palace_stems: *const [cycle_len]Stem,
) FiveElementBureau {
    return five_element_bureau.fromMingPalace(
        palace_stems[ming_palace.toInt()],
        ming_palace.toBranch(),
    );
}

/// 按寅起坐标生成十二宫名、支、干。
fn computePalacePlacements(
    ming_palace: YinIndex,
    palace_stems: *const [cycle_len]Stem,
) [cycle_len]PalacePlacement {
    var placements: [cycle_len]PalacePlacement = undefined;

    for (PalaceName.all) |name| {
        const yin = YinIndex.wrap(@as(i32, @intCast(ming_palace.toInt())) - @as(i32, @intCast(name.index())));
        placements[yin.toInt()] = .{
            .name = name,
            .branch = yin.toBranch(),
            .stem = palace_stems[yin.toInt()],
        };
    }
    return placements;
}

fn palaceNameAt(
    yin: YinIndex,
    placements: *const [cycle_len]PalacePlacement,
) PalaceName {
    return placements[yin.toInt()].name;
}

/// 一次完成十四主星与首批四颗辅星的落宫计算和十八槽组装。
fn computeStarBranches(
    bureau: FiveElementBureau,
    month: u8,
    day: u8,
    hour: u8,
) [StarName.count]Branch {
    const day_index = day - 1;
    const zi_wei_yin0 = zi_wei_yin0_by_bureau_day[bureauIndex(bureau)][day_index];
    var branches = major_branches_by_zi_wei_yin0[zi_wei_yin0];

    branches[star.index(.zuo_fu)] = zuo_you_by_month[month][0];
    branches[star.index(.you_bi)] = zuo_you_by_month[month][1];
    branches[star.index(.wen_chang)] = chang_qu_by_hour[hour][0];
    branches[star.index(.wen_qu)] = chang_qu_by_hour[hour][1];
    return branches;
}

fn buildPalaces(
    origin_palace: YinIndex,
    placements: *const [cycle_len]PalacePlacement,
    branches_by_star: *const [StarName.count]Branch,
) [cycle_len]Palace {
    const transformations = buildTransformations(origin_palace, placements, branches_by_star);
    const stars = buildStars(branches_by_star, &transformations);
    return assemblePalaces(placements, &stars, &transformations.by_palace);
}

const TransformationAssembly = struct {
    origin_by_star: [StarName.count]?Transformation,
    inward_by_star: [StarName.count]?Transformation,
    outward_by_star: [StarName.count]?Transformation,
    by_palace: [cycle_len][Transformation.all.len]palace.PalaceTransformation,
};

fn buildTransformations(
    origin_palace: YinIndex,
    placements: *const [cycle_len]PalacePlacement,
    branches_by_star: *const [StarName.count]Branch,
) TransformationAssembly {
    var result = TransformationAssembly{
        .origin_by_star = [_]?Transformation{null} ** StarName.count,
        .inward_by_star = [_]?Transformation{null} ** StarName.count,
        .outward_by_star = [_]?Transformation{null} ** StarName.count,
        .by_palace = undefined,
    };

    for (placements, 0..) |source, yin0| {
        const source_palace: YinIndex = @enumFromInt(yin0);
        for (Transformation.all, 0..) |current_transformation, transformation_index| {
            const star_name = star.fromStemTransformation(source.stem, current_transformation);
            const star_index = star.index(star_name);
            const target_branch = branches_by_star[star_index];
            const target_palace = YinIndex.fromBranch(target_branch);

            if (source_palace == origin_palace) {
                std.debug.assert(result.origin_by_star[star_index] == null);
                result.origin_by_star[star_index] = current_transformation;
            }
            if (source.branch == target_branch) {
                std.debug.assert(result.outward_by_star[star_index] == null);
                result.outward_by_star[star_index] = current_transformation;
            }
            if (primitive.oppositeBranch(source.branch) == target_branch) {
                std.debug.assert(result.inward_by_star[star_index] == null);
                result.inward_by_star[star_index] = current_transformation;
            }

            result.by_palace[yin0][transformation_index] = .{
                .source_name = source.name,
                .source_branch = source.branch,
                .transformation = current_transformation,
                .target_name = palaceNameAt(target_palace, placements),
                .target_branch = target_branch,
                .star_name = star_name,
            };
        }
    }
    return result;
}

fn buildStars(
    branches_by_star: *const [StarName.count]Branch,
    transformations: *const TransformationAssembly,
) [cycle_len]PalaceStars {
    var stars = [_]PalaceStars{PalaceStars.init()} ** cycle_len;
    for (StarName.all) |name| {
        const star_index = star.index(name);
        const palace_index = YinIndex.fromBranch(branches_by_star[star_index]).toInt();
        const value = star.init(
            name,
            transformations.origin_by_star[star_index],
            .{
                .inward = transformations.inward_by_star[star_index],
                .outward = transformations.outward_by_star[star_index],
            },
        );
        palace.appendStar(&stars[palace_index], value) catch
            @panic("supported star placement exceeds PalaceStars capacity");
    }
    return stars;
}

fn assemblePalaces(
    placements: *const [cycle_len]PalacePlacement,
    stars: *const [cycle_len]PalaceStars,
    transformations: *const [cycle_len][Transformation.all.len]palace.PalaceTransformation,
) [cycle_len]Palace {
    var palaces: [cycle_len]Palace = undefined;
    for (&palaces, placements, stars, transformations) |*value, palace_placement, palace_stars, palace_transformations| {
        value.* = .{
            .name = palace_placement.name,
            .branch = palace_placement.branch,
            .stem = palace_placement.stem,
            .stars = palace_stars,
            .transformations = palace_transformations,
        };
    }
    return palaces;
}

fn bureauIndex(bureau: FiveElementBureau) usize {
    inline for (std.enums.values(FiveElementBureau), 0..) |candidate, index_value| {
        if (candidate == bureau) return index_value;
    }
    unreachable;
}

fn buildZiWeiYin0ByBureauDay() [5][30]u4 {
    var table: [5][30]u4 = undefined;
    for (&table, std.enums.values(FiveElementBureau)) |*row, bureau| {
        for (row, 0..) |*value, day_index| {
            value.* = computeZiWeiYin0(@intCast(day_index + 1), bureau.number());
        }
    }
    return table;
}

fn computeZiWeiYin0(day: u8, bureau_number: u8) u4 {
    const day_value: i32 = day;
    const bureau_value: i32 = bureau_number;
    const ceiling_quotient = @divTrunc(day_value + bureau_value - 1, bureau_value);
    const shortfall = ceiling_quotient * bureau_value - day_value;
    const signed_shortfall = if (shortfall == 0)
        0
    else if (@mod(shortfall, 2) == 1)
        -shortfall
    else
        shortfall;
    return @intCast(@mod(
        ceiling_quotient - 1 + signed_shortfall,
        @as(i32, @intCast(cycle_len)),
    ));
}

fn buildMajorBranchesByZiWeiYin0() [cycle_len][StarName.count]Branch {
    @setEvalBranchQuota(5000);
    var table: [cycle_len][StarName.count]Branch = undefined;
    for (&table, 0..) |*row, zi_wei_yin0| {
        row.* = computeMajorBranchesFromZiWeiYin0(@intCast(zi_wei_yin0));
    }
    return table;
}

const StarOffset = struct {
    name: StarName,
    offset: i32,
};

fn computeMajorBranchesFromZiWeiYin0(zi_wei_yin0: u4) [StarName.count]Branch {
    const zi_wei_value: i32 = zi_wei_yin0;
    const tian_fu_value: i32 = @intCast(@mod(-zi_wei_value, @as(i32, @intCast(cycle_len))));
    var branches = [_]Branch{.zi} ** StarName.count;
    const zi_wei_series = [_]StarOffset{
        .{ .name = .zi_wei, .offset = 0 },
        .{ .name = .tian_ji, .offset = 1 },
        .{ .name = .tai_yang, .offset = 3 },
        .{ .name = .wu_qu, .offset = 4 },
        .{ .name = .tian_tong, .offset = 5 },
        .{ .name = .lian_zhen, .offset = 8 },
    };
    const tian_fu_series = [_]StarOffset{
        .{ .name = .tian_fu, .offset = 0 },
        .{ .name = .tai_yin, .offset = 1 },
        .{ .name = .tan_lang, .offset = 2 },
        .{ .name = .ju_men, .offset = 3 },
        .{ .name = .tian_xiang, .offset = 4 },
        .{ .name = .tian_liang, .offset = 5 },
        .{ .name = .qi_sha, .offset = 6 },
        .{ .name = .po_jun, .offset = 10 },
    };

    for (zi_wei_series) |entry| {
        setStarBranch(&branches, entry.name, YinIndex.wrap(zi_wei_value - entry.offset));
    }
    for (tian_fu_series) |entry| {
        setStarBranch(&branches, entry.name, YinIndex.wrap(tian_fu_value + entry.offset));
    }
    return branches;
}

fn buildZuoYouByMonth() [cycle_len][2]Branch {
    var table: [cycle_len][2]Branch = undefined;
    for (&table, 0..) |*row, month| {
        row.* = .{
            YinIndex.wrap(2 + @as(i32, @intCast(month))).toBranch(),
            YinIndex.wrap(8 - @as(i32, @intCast(month))).toBranch(),
        };
    }
    return table;
}

fn buildChangQuByHour() [cycle_len][2]Branch {
    var table: [cycle_len][2]Branch = undefined;
    for (&table, 0..) |*row, hour| {
        row.* = .{
            YinIndex.wrap(8 - @as(i32, @intCast(hour))).toBranch(),
            YinIndex.wrap(2 + @as(i32, @intCast(hour))).toBranch(),
        };
    }
    return table;
}

fn setStarBranch(branches: *[StarName.count]Branch, name: StarName, yin: YinIndex) void {
    branches[star.index(name)] = yin.toBranch();
}

test "安星预计算表与生成公式在全部支持输入上一致" {
    const bureaus = [_]FiveElementBureau{
        .water_two,
        .wood_three,
        .metal_four,
        .earth_five,
        .fire_six,
    };

    for (bureaus) |bureau| {
        for (1..31) |day| {
            for (0..Branch.all.len) |month| {
                for (0..Branch.all.len) |hour| {
                    const actual = computeStarBranches(
                        bureau,
                        @intCast(month),
                        @intCast(day),
                        @intCast(hour),
                    );
                    const zi_wei_yin0 = computeZiWeiYin0(@intCast(day), bureau.number());
                    var expected = computeMajorBranchesFromZiWeiYin0(zi_wei_yin0);
                    expected[star.index(.zuo_fu)] = YinIndex.wrap(2 + @as(i32, @intCast(month))).toBranch();
                    expected[star.index(.you_bi)] = YinIndex.wrap(8 - @as(i32, @intCast(month))).toBranch();
                    expected[star.index(.wen_chang)] = YinIndex.wrap(8 - @as(i32, @intCast(hour))).toBranch();
                    expected[star.index(.wen_qu)] = YinIndex.wrap(2 + @as(i32, @intCast(hour))).toBranch();
                    try std.testing.expectEqual(expected, actual);

                    var stars_per_branch = [_]u8{0} ** Branch.all.len;
                    for (actual) |branch| stars_per_branch[branch.index()] += 1;
                    for (stars_per_branch) |count| {
                        try std.testing.expect(count <= 6);
                    }
                }
            }
        }
    }
}
